src/mbr.c

```diff
@@ -0,0 +1,425 @@
+
+	if (ioctl(fd, BLKGETSIZE, sectors) != 0) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"ioctl command 0x%04x failed: %s",
+				BLKGETSIZE, g_strerror(errno));
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean get_hd_geometry(gint fd, struct hd_geometry *geometry,
+		GError **error)
+{
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+	g_assert_nonnull(geometry);
+
+	if (ioctl(fd, HDIO_GETGEO, geometry) != 0) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"ioctl command 0x%04x failed: %s",
+				HDIO_GETGEO, g_strerror(errno));
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean validate_region(gint fd, guint64 start, guint64 size,
+		guint sector_size, GError **error)
+{
+	gboolean res = FALSE;
+	guint number_of_sectors;
+	GError *ierror = NULL;
+
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+	if (start < sizeof(struct mbr) || size == 0) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"no valid region configuration");
+		goto out;
+	}
+
+	if ((start % (2 * sector_size)) != 0 || (size % (2 * sector_size)) != 0) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"region configuration is not aligned to two times the sector-size (%d)",
+				2 * sector_size);
+		goto out;
+	}
+
+	res = get_number_of_sectors(fd, &number_of_sectors, &ierror);
+	if (!res) {
+		g_propagate_error(error, ierror);
+		goto out;
+	}
+
+	if ((start + size) >= (guint64)number_of_sectors * sector_size) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"region configuration is bigger than device");
+		res = FALSE;
+		goto out;
+	}
+
+out:
+	return res;
+}
+
+static gboolean read_mbr(gint fd, struct mbr *mbr, GError **error)
+{
+	gboolean res = FALSE;
+
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+	if (read(fd, mbr, sizeof(*mbr)) != sizeof(*mbr)) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"read: %s", g_strerror(errno));
+		goto out;
+	}
+
+	if (mbr->magic_number[0] != MBR_MAGIC_NUMBER_L ||
+	    mbr->magic_number[1] != MBR_MAGIC_NUMBER_H) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"no valid master boot record found");
+		goto out;
+	}
+
+	res = TRUE;
+out:
+	return res;
+}
+
+static gboolean is_region_free(guint64 region_start, guint64 region_size,
+		const struct partition_tbl_entry *partition_tbl, GError **error)
+{
+	guint32 entry_end;
+	guint i;
+
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+	for (i = 0; i < MBR_NUMBER_OF_PARTITIONS; i++) {
+		if (i == BOOT_PARTITION_ENTRY)
+			continue;
+
+		entry_end = partition_tbl[i].partition_start +
+		            partition_tbl[i].partition_size - 1;
+
+		if (region_start >= partition_tbl[i].partition_start &&
+		    region_start <= entry_end)
+			break;
+
+		if (partition_tbl[i].partition_start >= region_start &&
+		    partition_tbl[i].partition_start <=
+		    region_start + region_size - 1)
```

Comment:
Same as above. If you increase `region_size`, the update shall fail, because the region overlaps with the next partition, but happily results in an invalid partition table.